core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/unmarshall/XmlResponseParserUtils.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.protocols.xml.internal.unmarshall;
+
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.SdkPojo;
+import software.amazon.awssdk.core.protocol.MarshallLocation;
+import software.amazon.awssdk.http.SdkHttpFullResponse;
+import software.amazon.awssdk.protocols.query.unmarshall.XmlDomParser;
+import software.amazon.awssdk.protocols.query.unmarshall.XmlElement;
+
+/**
+ * Static methods to assist with parsing the response of AWS XML requests.
+ */
+@SdkInternalApi
+public final class XmlResponseParserUtils {
+    private XmlResponseParserUtils() {
+    }
+
+    /**
+     * Parse an XML response if one is expected and available. If we are not expecting a payload, but the HTTP response
+     * code shows an error then we will parse it anyway, as it should contain a serialized error.
+     * @param sdkPojo the SDK builder object associated with the final response
+     * @param response the HTTP response
+     * @return A parsed XML document or an empty XML document if no payload/contents were found in the response.
+     */
+    public static XmlElement parse(SdkPojo sdkPojo, SdkHttpFullResponse response) {
+
+        try {
+            if ((!response.isSuccessful() || hasPayloadMembers(sdkPojo)) && response.content().isPresent()) {
```

Comment:
Can we extract the check statement to a well-named, well-documented separate method? It's a bit unclear to me that why just checking `response.content().isPresent` is not sufficient.

Reply:
The reason is that sometimes the response content is present but it's like an empty stream and the XML parser gets called anyway and then throws an exception because it's not valid XML so I assume this logic was added to avoid unecessary parsing of an empty stream. This code was refactored from another place, it is not new. I'll add a comment.